server/mqtt.go

```diff
@@ -2297,12 +2375,6 @@ func (as *mqttAccountSessionManager) processSubs(sess *mqttSession, c *client,
 		sess.cons[sid] = cc
 	}
 
-	serializeRMS := func(sub *subscription) {
```

Comment:
So, the act of serializing the header depends on the sub (QOS), and requires the session lock (PIs).

That's why I kept it where it was, but I removed the payload from there, we now just move the []byte pointer.

Like I mentioned in my other comment, we scan the RM "subs" (as.cl) again, that can be cached; otherwise I am not sure what else is there to do. The PI allocation must be done under the session lock, of course.

Reply:
(resolving this in favor of https://github.com/nats-io/nats-server/pull/4810/files?diff=split&w=1#r1406463798)